File: sdk/swift-sdk/Sources/OwnFirebaseSDK/Storage.swift
import Foundation

public class StorageService: OwnFirebaseClient {
  // MARK: - Upload URL Management

  public func getUploadUrl(
    path: String,
    contentType: String = "application/octet-stream",
    size: Int? = nil,
    metadata: [String: AnyCodable]? = nil
  ) async throws -> StorageUploadUrl {
    let body = GetUploadUrlRequest(
      path: path,
      content_type: contentType,
      size: size,
      metadata: metadata
    )
    return try await request(
      "POST",
      url: projectUrl("storage/upload-url/"),
      body: body
    )
  }

  public func confirmUpload(fileId: String) async throws -> StorageObject {
    let body = ConfirmUploadRequest(file_id: fileId)
    return try await request(
      "POST",
      url: projectUrl("storage/confirm/"),
      body: body
    )
  }

  // MARK: - File Operations

  public func listFiles(prefix: String? = nil) async throws -> PaginatedResponse<StorageObject> {
    var options = RequestOptions()
    if let prefix = prefix {
      options.query = ["prefix": prefix]
    }

    return try await request(
      "GET",
      url: projectUrl("storage/files/"),
      options: options
    )
  }

  public func getFile(path: String) async throws -> StorageObject {
    return try await request(
      "GET",
      url: projectUrl("storage/files/\(path)/")
    )
  }

  public func deleteFile(path: String) async throws {
    try await requestVoid(
      "DELETE",
      url: projectUrl("storage/files/\(path)/")
    )
  }

  // MARK: - High-Level Upload Helper

  public func upload(
    data: Data,
    path: String,
    contentType: String = "application/octet-stream",
    metadata: [String: AnyCodable]? = nil
  ) async throws -> StorageObject {
    // Get upload URL
    let uploadUrlResponse = try await getUploadUrl(
      path: path,
      contentType: contentType,
      size: data.count,
      metadata: metadata
    )

    // Upload file to presigned URL
    try await uploadToPresignedUrl(
      url: uploadUrlResponse.upload_url,
      data: data,
      contentType: contentType
    )

    // Confirm upload
    return try await confirmUpload(fileId: uploadUrlResponse.file_id)
  }

  private func uploadToPresignedUrl(
    url: String,
    data: Data,
    contentType: String
  ) async throws {
    guard let presignedUrl = URL(string: url) else {
      throw OwnFirebaseError.invalidURL
    }

    var request = URLRequest(url: presignedUrl)
    request.httpMethod = "PUT"
    request.setValue(contentType, forHTTPHeaderField: "Content-Type")
    request.httpBody = data

    let (_, response) = try await URLSession.shared.data(for: request)

    guard let httpResponse = response as? HTTPURLResponse else {
      throw OwnFirebaseError.invalidResponse
    }

    guard httpResponse.statusCode >= 200 && httpResponse.statusCode < 300 else {
      throw OwnFirebaseError.apiError(
        APIError(
          status: httpResponse.statusCode,
          message: HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode),
          detail: nil
        )
      )
    }
  }

  // MARK: - Download Helper

  public func downloadFile(url: String) async throws -> Data {
    guard let fileUrl = URL(string: url) else {
      throw OwnFirebaseError.invalidURL
    }

    let (data, response) = try await URLSession.shared.data(from: fileUrl)

    guard let httpResponse = response as? HTTPURLResponse else {
      throw OwnFirebaseError.invalidResponse
    }

    guard httpResponse.statusCode >= 200 && httpResponse.statusCode < 300 else {
      throw OwnFirebaseError.apiError(
        APIError(
          status: httpResponse.statusCode,
          message: HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode),
          detail: nil
        )
      )
    }

    return data
  }
}

// MARK: - Request Types

private struct GetUploadUrlRequest: Encodable {
  let path: String
  let content_type: String
  let size: Int?
  let metadata: [String: AnyCodable]?
}

private struct ConfirmUploadRequest: Encodable {
  let file_id: String
}

File: sdk/src/storage.ts
import { OwnFirebaseClient } from './client';
import type { StorageObject, StorageUploadUrl, PaginatedResponse } from './types';

export class StorageSDK extends OwnFirebaseClient {
  /**
   * Request a presigned upload URL from MinIO/S3 for direct browser upload.
   */
  async getUploadUrl(options: {
    filename: string;
    contentType: string;
    path?: string;
  }): Promise<StorageUploadUrl> {
    return this.request('POST', this.projectUrl('storage/upload-url/'), {
      filename: options.filename,
      content_type: options.contentType,
      path: options.path,
    });
  }

  /**
   * Confirm a direct upload after the client has PUT to the presigned URL.
   */
  async confirmUpload(objectKey: string): Promise<StorageObject> {
    return this.request('POST', this.projectUrl('storage/confirm/'), {
      object_key: objectKey,
    });
  }

  async listFiles(prefix?: string): Promise<PaginatedResponse<StorageObject>> {
    const query: Record<string, string> = {};
    if (prefix) query['prefix'] = prefix;
    return this.request(
      'GET',
      this.projectUrl('storage/files/'),
      undefined,
      { query }
    );
  }

  async getFile(path: string): Promise<StorageObject> {
    return this.request('GET', this.projectUrl(`storage/files/${path}/`));
  }

  async deleteFile(path: string): Promise<void> {
    return this.request('DELETE', this.projectUrl(`storage/files/${path}/`));
  }

  /**
   * High-level helper: request an upload URL, PUT the file, then confirm.
   * Only works in environments with fetch + Blob support (browsers, Deno, Node 18+).
   */
  async upload(
    file: Blob | Buffer | ArrayBuffer,
    options: { filename: string; contentType: string; path?: string }
  ): Promise<StorageObject> {
    const { upload_url, object_key } = await this.getUploadUrl(options);

    const putResponse = await fetch(upload_url, {
      method: 'PUT',
      headers: { 'Content-Type': options.contentType },
      body: file as any,
    });

    if (!putResponse.ok) {
      throw new Error(
        `Upload to presigned URL failed: ${putResponse.status} ${putResponse.statusText}`
      );
    }

    return this.confirmUpload(object_key);
  }
}
